src/libs/sfdk/dockercontainer.cpp

```diff
@@ -0,0 +1,341 @@
+           QObject::connect(runner.get(), &ProcessRunner::failure, Sdk::instance(), [=]() {
+               SdkPrivate::commandQueue()->cancelBatch(batch);
+               callIf(context_, functor, false);
+           });
+           ProcessRunner *runner_ = runner.get();
+           SdkPrivate::commandQueue()->enqueue(std::move(runner));
+           return runner_;
+       };
+
+    CommandQueue::BatchId batch = SdkPrivate::commandQueue()->beginBatch();
+
+    QStringList arguments;
+    arguments.append(QLatin1String(STOP));
+    arguments.append(q->m_containerId);
+
+    enqueue(arguments, batch);
+
+    arguments[0] = QLatin1String(COMMIT);
+    arguments.append(q->name());
+
+    enqueue(arguments, batch);
+
+    SdkPrivate::commandQueue()->enqueueCheckPoint(context,  [=]() { functor(true); });
+    SdkPrivate::commandQueue()->endBatch();
+}
+
+void DockerContainerPrivate::probe(const QObject *context, const Functor<VirtualMachinePrivate::BasicState, bool> &functor) const
+{
+    Q_Q(const DockerContainer);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    DockerManager::fetchRegisteredVirtualMachines(q,
+        [=](const QStringList &registeredVms, bool ok) {
+            if (!ok) {
+                VirtualMachinePrivate::BasicState state;
+                functor(state, false);
+            } else {
+                if (!registeredVms.contains(q->name())) {
+                    VirtualMachinePrivate::BasicState state;
+                    functor(state, true);
+                }
+                QStringList arguments;
+                arguments.append(QLatin1String(CONTAINER_LIST));
+                arguments.append(QLatin1String(FILTER_PREFIX)+q->name());
+                arguments.append(QLatin1String(FORMAT));
+                auto runner = std::make_unique<ProcessRunner>(DockerManager::dockerPath(), arguments);
+                QObject::connect(runner->process(), QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished), context,
+                        [=, runner = runner.get()](int exitCode, QProcess::ExitStatus exitStatus) {
+                    VirtualMachinePrivate::BasicState state = VirtualMachinePrivate::Existing;
+                    if (exitStatus != QProcess::NormalExit) {
+                        functor(state, false);
+                        return;
+                    }
+                    if (exitCode != 0) {
+                        functor(state, false);
+                        return;
+                    }
+
+                    QString runningContainers = QString::fromLocal8Bit(runner->process()->readAllStandardOutput());
+                    bool isRunning = runningContainers.contains(q->name());
+                    if (isRunning) {
+                        state |= (VirtualMachinePrivate::Running | VirtualMachinePrivate::Headless);
+                        if (q->m_containerId.isEmpty())
+                            q->m_containerId = runningContainers.split(' ')[1].trimmed();
```

Comment:
One never knows.. check for required capacity?

Second, not that it should make any harm to performance here, but I guess `operator[]` detaches unnecessarily even in this context - I would simply avoid its use for reading.

Reply:
What do you mean by required capacity?

Reply:
I mean to check that `runningContainers.split(' ')` has at least two elements. Out of bound access results in abortion, so I would prefer checking that.

Reply:
Right. Thanks. Fixed now - if the output contains something else than what's expected, it will print an error message and assume that the container is not running.

Reply:
I would prefer using just `QTC_ASSERT(nameAndId.count() == 2 && nameAndId.first() == q->name(), continue)` there.

Reply:
makes sense - fixed